docker crashes when using gelf log driver in tcp mode and logging server becomes unavailable

**Description**

I experimented with new tcp mode for gelf log driver and found out that docker crashes when log server goes down.

**Steps to reproduce the issue:**
1. start something to imitate log server ex: "nc -l -p 12000"
2. start container "docker run  --rm -it --log-driver gelf --log-opt gelf-address=tcp://127.0.0.1:12000  debian bash"
3. type something in the container's terminal (generate some logs)
4. stop the nc process
5. type something again in the container's terminal (generate some more logs)

**Describe the results you received:**
Docker crashes with the following logs:
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x58 pc=0x2455211]

goroutine 210 [running]:
github.com/docker/docker/vendor/github.com/Graylog2/go-gelf/gelf.(*TCPWriter).writeToSocketWithReconnectAttempts(0xc421d57490, 0xc421d59800, 0x1b5, 0x400, 0x0, 0x0, 0xc4205d0d40)
	/var/tmp/portage/app-emulation/docker-17.11.0/work/docker-17.11.0/src/github.com/docker/docker/vendor/github.com/Graylog2/go-gelf/gelf/tcpwriter.go:83 +0x71
github.com/docker/docker/vendor/github.com/Graylog2/go-gelf/gelf.(*TCPWriter).WriteMessage(0xc421d57490, 0xc421b29200, 0x1d, 0x100)
	/var/tmp/portage/app-emulation/docker-17.11.0/work/docker-17.11.0/src/github.com/docker/docker/vendor/github.com/Graylog2/go-gelf/gelf/tcpwriter.go:53 +0x92
github.com/docker/docker/daemon/logger/gelf.(*gelfLogger).Log(0xc4204b8700, 0xc421d1c0e0, 0x1d, 0x4000)
	/var/tmp/portage/app-emulation/docker-17.11.0/work/docker-17.11.0/src/github.com/docker/docker/daemon/logger/gelf/gelf.go:186 +0x255
github.com/docker/docker/daemon/logger.(*Copier).copySrc(0xc421c4fc40, 0x2a462be, 0x6, 0x7f754d89dc10, 0xc421c4fa40)
	/var/tmp/portage/app-emulation/docker-17.11.0/work/docker-17.11.0/src/github.com/docker/docker/daemon/logger/copier.go:100 +0x471
created by github.com/docker/docker/daemon/logger.(*Copier).Run
	/var/tmp/portage/app-emulation/docker-17.11.0/work/docker-17.11.0/src/github.com/docker/docker/daemon/logger/copier.go:46 +0x118
```

**Describe the results you expected:**
No crash.

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `docker version`:**

```
Client:
 Version:      17.11.0-ce
 API version:  1.34
 Go version:   go1.8.3
 Git commit:   1caf76c
 Built:        Mon Nov 27 10:56:33 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.11.0-ce
 API version:  1.34 (minimum version 1.12)
 Go version:   go1.8.3
 Git commit:   1caf76c
 Built:        Mon Nov 27 10:55:29 2017
 OS/Arch:      linux/amd64
 Experimental: false
```

**Output of `docker info`:**

```
Containers: 4
 Running: 4
 Paused: 0
 Stopped: 0
Images: 747
Server Version: 17.11.0-ce
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 992280 (expected: 992280e8e265f491f7a624ab82f3e238be086e49)
runc version: 0351df1 (expected: 0351df1c5a66838d0c392b4ac4cf9450de844e2d)
init version: v0.16.1 (expected: 949e6facb77383876aeff8a6944dde66b3089574)
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.9.16-gentoo
Operating System: Gentoo/Linux
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 7.673GiB
Name: stas.office
ID: 6XIR:QBFD:VPGE:A6QA:6EIA:OOE4:CWZB:4ODI:IOZF:THCT:HOP4:AYXS
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 18
 Goroutines: 34
 System Time: 2017-11-27T20:31:26.54513255+05:00
 EventsListeners: 0
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
physical

--- a/vendor.conf
+++ b/vendor.conf
@@ -77,7 +77,7 @@ github.com/syndtr/gocapability 2c00daeb6c3b45114c80ac44119e7b8801fdd852
 github.com/golang/protobuf 7a211bcf3bce0e3f1d74f9894916e6f116ae83b4
 
 # gelf logging driver deps
-github.com/Graylog2/go-gelf v2
+github.com/Graylog2/go-gelf ba920adcee0319adcc15f52851f1458f56547975
 
 github.com/fluent/fluent-logger-golang v1.3.0
 # fluent-logger-golang deps

--- a/vendor.conf
+++ b/vendor.conf
@@ -77,7 +77,7 @@ github.com/syndtr/gocapability 2c00daeb6c3b45114c80ac44119e7b8801fdd852
 github.com/golang/protobuf 7a211bcf3bce0e3f1d74f9894916e6f116ae83b4
 
 # gelf logging driver deps
-github.com/Graylog2/go-gelf ba920adcee0319adcc15f52851f1458f56547975
+github.com/Graylog2/go-gelf 4143646226541087117ff2f83334ea48b3201841
 
 github.com/fluent/fluent-logger-golang v1.3.0
 # fluent-logger-golang deps

